Sources/AsyncHTTPClient/Utils.swift

```diff
@@ -47,21 +53,85 @@ public final class HTTPClientCopyingDelegate: HTTPClientResponseDelegate {
 }
 
 extension ClientBootstrap {
-    static func makeHTTPClientBootstrapBase(group: EventLoopGroup, host: String, port: Int, configuration: HTTPClient.Configuration, channelInitializer: ((Channel) -> EventLoopFuture<Void>)? = nil) -> ClientBootstrap {
-        return ClientBootstrap(group: group)
-            .channelOption(ChannelOptions.socket(SocketOptionLevel(IPPROTO_TCP), TCP_NODELAY), value: 1)
+    fileprivate func makeClientTCPBootstrap(
+        host: String,
+        requiresTLS: Bool,
+        configuration: HTTPClient.Configuration
+    ) throws -> NIOClientTCPBootstrap {
+        let tlsConfiguration = configuration.tlsConfiguration ?? TLSConfiguration.forClient()
+        let sslContext = try NIOSSLContext(configuration: tlsConfiguration)
+        let hostname = (!requiresTLS || host.isIPAddress) ? nil : host
+        let tlsProvider = try NIOSSLClientTLSProvider<ClientBootstrap>(context: sslContext, serverHostname: hostname)
+        return NIOClientTCPBootstrap(self, tls: tlsProvider)
+    }
+}
 
+extension NIOClientTCPBootstrap {
+    
+    /// create a TCP Bootstrap based off what type of `EventLoop` has been passed to the function.
+    fileprivate static func makeBootstrap(
+        on eventLoop: EventLoop,
+        host: String,
+        requiresTLS: Bool,
+        configuration: HTTPClient.Configuration
+    ) throws -> NIOClientTCPBootstrap {
+        let bootstrap: NIOClientTCPBootstrap
+        #if canImport(Network)
+        // if eventLoop is compatible with NIOTransportServices create a NIOTSConnectionBootstrap
+        if #available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *),
+            var tsBootstrap = NIOTSConnectionBootstrap(validatingGroup: eventLoop) {
+            
+            tsBootstrap = tsBootstrap.channelOption(NIOTSChannelOptions.waitForActivity, value: false)
+            let tlsConfiguration = configuration.tlsConfiguration ?? TLSConfiguration.forClient()
+            // if we have a proxy and require TLS then use NIOSSL tls support
+            if configuration.proxy != nil, requiresTLS {
+                let sslContext = try NIOSSLContext(configuration: tlsConfiguration)
+                let hostname = (!requiresTLS || host.isIPAddress) ? nil : host
+                bootstrap = try NIOClientTCPBootstrap(tsBootstrap, tls: NIOSSLClientTLSProvider(context: sslContext, serverHostname: hostname))
```

Comment:
Why are we adding a `TLSProvider` here if we're not enabling it? Proxy support doesn't use the `TLSProvider` interface so far as I can see.

Reply:
This was suggested by @weissi and @artemredkin. For proxies if needed we enable the TLS at a later point using `channel.pipeline.addSSLHandlerIfNeeded`. Although we maybe creating two `NIOSSLContext`, one to place in the `NIOClientTCPBootstrap` and then the one that is actually used. @weissei Should I replace the one in `NIOClientTCPBootstrap` with a `NIOInsecureNoTLS`

Reply:
I think the need to add the later TLS in `addSSLHandlerIfNeeded` is fine, but we can't use the abstract bootstrap protocol to do it, so it seems unnecessary to add a TLS provider here.

Reply:
Yes you are correct

Reply:
When running a proxy, I've setup the `NIOClientTCPBootstrap` with `tls` of `NIOInsecureNoTLS`